DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/README.md

```diff
@@ -0,0 +1,250 @@
+
+ac_int automatically promotes the result of all operations to the number of bits needed to represent all possible results without overflowing. For example, the addition of two 8-bit integers results in a 9-bit result to support overflow.
+
+However, if the user attempts to store the result in an 8-bit container, ac_int will let the user do this, but this leads to the discard of the extra carry bit. The responsibility lies on the user to use the correct datatype.
+
+These promotions rules are consistent across all architectures so the behavior should be equivalent on x86 or on FPGA.
+
+#### Shift Operation
+
+The behavior of a shift operation with ac_ints is slightly different from its behvior with native integer types. For full details, see the ac_int documentation in the file `ac_data_types_ref.pdf`. Some key points to remember are as follows:
+- If the datatype of the shift amount is signed, negative shift amounts are accepted. A shift by a negative amount is equivalent to a positive shift in the opposite direction.
+- If the user knows that the shift direction and that the shift value is always positive, it is advisable to use an unsigned datatype for the shift value to obtain better QoR.
+- Shift values of greater than the width of the datatypes are treated as a shift equal to the width of the datatype.
+
+Further, the operation can be done more efficiently by specifying the amount to shift with the smallest possible ac_int.
+
+For example, in the tutorial, two kernels perform the left shift operation: `ShiftLeft` and `EfficientShiftLeft`. Both operate on an 14 bits wide ac_int. The former stores the shift amount in an ac_int which is 14 bits wide and the latter stores it in an ac_int which is 4 bits wide. The latter will generate lesser hardware.
+
+#### Bit Slice Operations
+
+The kernels `GetBitSlice` and `SetBitSlice` show how to read from and write to specific bits of an ac_int number. Note that only static bit widths are supported with such "slice" operations.
+
+For detailed documentation on the `set_slc` and `slc` APIs please see the file `ac_data_types_ref.pdf`
+
+## Key Concepts
+* The ac_int data-type can be used to generate hardware for only as many bits as is needed by the operation as compared to native integer types which generate hardware for the entire type width.
+* The left shift operation on ac_int can be implemented more efficiently when the amount to shift with is stored in a minimally sized ac_int.
+* The ac_int data-type offers functions for several useful operations including reading and writing of certain bits of an ac_int number. This can be very useful in creating bit masks.
+
+## License
+
+Code samples are licensed under the MIT license. See
+[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.
```

Comment:
do we need a license for the ac_data_types_ref.pdf?

Reply:
Let me follow up on this.